packages/agreement-platformstate-writer/src/consumerServiceV1.ts

```diff
@@ -1,24 +1,351 @@
+): Promise<void> => {
+  const primaryKey = makePlatformStatesAgreementPK(agreement.id);
+
+  const existingAgreementEntry = await readAgreementEntry(
+    primaryKey,
+    dynamoDBClient
+  );
+  const GSIPK_consumerId_eserviceId = makeGSIPKConsumerIdEServiceId({
+    consumerId: agreement.consumerId,
+    eserviceId: agreement.eserviceId,
+  });
+
+  if (existingAgreementEntry) {
+    if (existingAgreementEntry.version > incomingVersion) {
+      // Stops processing if the message is older than the agreement entry
+      return Promise.resolve();
+    } else {
+      await updateAgreementStateInPlatformStatesEntry(
+        dynamoDBClient,
+        primaryKey,
+        agreementStateToItemState(agreement.state),
+        incomingVersion
+      );
+    }
+  }
+
+  const pkCatalogEntry = makePlatformStatesEServiceDescriptorPK({
+    eserviceId: agreement.eserviceId,
+    descriptorId: agreement.descriptorId,
+  });
+  const catalogEntry = await readCatalogEntry(pkCatalogEntry, dynamoDBClient);
+
+  const GSIPK_eserviceId_descriptorId = makeGSIPKEServiceIdDescriptorId({
+    eserviceId: agreement.eserviceId,
+    descriptorId: agreement.descriptorId,
+  });
+
+  if (
+    await isAgreementTheLatest(
+      GSIPK_consumerId_eserviceId,
+      agreement.id,
+      dynamoDBClient
+    )
+  ) {
+    // token-generation-states
+    await updateAgreementStateAndDescriptorInfoOnTokenStates({
+      GSIPK_consumerId_eserviceId,
+      agreementState: agreement.state,
+      dynamoDBClient,
+      GSIPK_eserviceId_descriptorId,
+      catalogEntry,
+    });
+  }
+};
+
+const handleArchiving = async (
+  agreement: Agreement,
+  dynamoDBClient: DynamoDBClient
+): Promise<void> => {
+  const primaryKey = makePlatformStatesAgreementPK(agreement.id);
+  const GSIPK_consumerId_eserviceId = makeGSIPKConsumerIdEServiceId({
+    consumerId: agreement.consumerId,
+    eserviceId: agreement.eserviceId,
+  });
+
+  if (
+    await isAgreementTheLatest(
+      GSIPK_consumerId_eserviceId,
+      agreement.id,
+      dynamoDBClient
+    )
+  ) {
+    // token-generation-states only if agreement is the latest
+
+    await updateAgreementStateOnTokenStates({
+      GSIPK_consumerId_eserviceId,
+      agreementState: agreement.state,
+      dynamoDBClient,
+    });
+  }
+
+  await deleteAgreementEntry(primaryKey, dynamoDBClient);
+};
+
+const handleUpgrade = async (
+  agreement: Agreement,
+  dynamoDBClient: DynamoDBClient,
+  msgVersion: number
+): Promise<void> => {
+  const primaryKey = makePlatformStatesAgreementPK(agreement.id);
+  const agreementEntry = await readAgreementEntry(primaryKey, dynamoDBClient);
+
+  const pkCatalogEntry = makePlatformStatesEServiceDescriptorPK({
+    eserviceId: agreement.eserviceId,
+    descriptorId: agreement.descriptorId,
+  });
+  const catalogEntry = await readCatalogEntry(pkCatalogEntry, dynamoDBClient);
+  if (!catalogEntry) {
+    // TODO double-check
+    throw genericInternalError("Catalog entry not found");
+  }
```

Comment:
To do: double-check

Reply:
[same](https://github.com/pagopa/interop-be-monorepo/pull/987#discussion_r1818914641)